### ISSUE TITLE
Influence of number of GPUs used during training

### ISSUE THREAD

**Issue by mprzewie** [facebookresearch/dinov2#299]
Hi, 

I am trying to reproduce your results on ImageNet-1k. I followed the instructions in the README with one exception - I trained on a single A100 node (8x NVIDIA A100-SXM4-40GB) instead of four, due to lack of such computational resources. If I understand the logs correctly, training was performed for 125000 iterations and finished after a day without any errors.

Next, I ran the following command:

```
PYTHONPATH=. python dinov2/run/eval/knn.py   --gpus 1  --config-file $OUTPUT_DIR/config.yaml     \
--pretrained-weights $OUTPUT_DIR/eval/training_112499/teacher_checkpoint.pth     --output-dir $OUTPUT_DIR/eval/training_112499/knn    \
 --train-dataset ImageNet:split=TRAIN:root=/net/people/plgrid/plgprzewiez/storage/datasets/ImageNet/ILSVRC/Data/CLS-LOC:extra=/net/people/plgrid/plgprzewiez/storage/datasets/ImageNet/ILSVRC/Data/DINOV2_extra/    \
  --val-dataset ImageNet:split=VAL:root=/net/people/plgrid/plgprzewiez/storage/datasets/ImageNet/ILSVRC/Data/CLS-LOC:extra=/net/people/plgrid/plgprzewiez/storage/datasets/ImageNet/ILSVRC/Data/DINOV2_extra/
```

and obtained the following KNN evaluation results, which are a bit lower than 81.6%:

```
I20231105 09:26:14 895116 dinov2 knn.py:367] ('full', 10) classifier result: Top1: 77.35 Top5: 90.00
I20231105 09:26:14 895116 dinov2 knn.py:367] ('full', 20) classifier result: Top1: 77.10 Top5: 91.46
I20231105 09:26:14 895116 dinov2 knn.py:367] ('full', 100) classifier result: Top1: 75.56 Top5: 93.15
I20231105 09:26:14 895116 dinov2 knn.py:367] ('full', 200) classifier result: Top1: 74.69 Top5: 93.32
```

Could the fact that I had used fewer GPUs for training have resulted in lower metrics?

**Comment by GravityZL** [comment]
> Yes, using fewer GPUs means the total batch size for training is smaller, so the results are not as good (also because the optimal hyperparameters may be slightly different with a smaller batch size).

Then if we increase the iterations, could we meet the same performance with fewer gpus?

**Comment by qasfb** [comment]
Same performance I'm not sure, but longer training should help. There might be slight differences in final performances because of the hyperparameters for your batch size.

https://github.com/facebookresearch/dinov2/blob/main/dinov2/configs/ssl_default_config.yaml#L68

The OFFICIAL EPOCH LENGTH is the parameter you want to change, it defines the number of iterations for one "epoch".
If you multiply it by 2, the training will be twice longer, with similar (dilated) schedules.

**Comment by mprzewie** [comment]
Thank you!

**Comment by mprzewie** [comment]
On a side note, How do we interpret the `N` in ('full', `N`) from the logs?

**Comment by qasfb** [comment]
So with 95% certainty I think this is using the N closest neighbors (with some hardcoded temperature parameter to increase the weight of closer neighbors) to determine the class prediction. In general you can just stick to the 10-neighbor prediction.

**Comment by mprzewie** [comment]
Makes sense! Thanks for both anwers!

**Comment by ripperthejack** [comment]
@mprzewie can you provide the config of your reproduce。 I would like to know the learning rate, batch size, and model type（small? base? large?） of your experiment

**Comment by ripperthejack** [comment]
@mprzewie in my experiment，the max batch size of an A100 40GB is 4 , what is the per_gpu_batch_size？

**Comment by ripperthejack** [comment]
@mprzewie  in my experiment，the max batch size of an A100 40GB is 4 , what is the per_gpu_batch_size  of yours？

**Comment by ripperthejack** [comment]
@mprzewie Hi , I would like to know if you have used ssl_defaulteconfig.yaml and vitl16_short.yaml as config files. and batch size is 64*8?  Did you use the official distributed module（submitit）？ in my case ，I use torchrun instead ，But training on 8 A100 40G using official configuration files（ssl_defaulteconfig.yaml and vitl16_short.yaml 64*8 batch size） takes 4 to 5 days。

can you provide more detail of your training?

**Comment by ripperthejack** [comment]
@mprzewie Hi , I would like to know if you have used ssl_defaulteconfig.yaml and vitl16_short.yaml as config files. and batch size is 64**8? Did you use the official distributed module（submitit）？ in my case ，I use torchrun instead ，But training on 8 A100 40G using official configuration files（ssl_defaulteconfig.yaml and vitl16_short.yaml 64**8 batch size） takes 4 to 5 days。

can you provide more detail of your training?

**Comment by ripperthejack** [comment]
@mprzewie  Hi，I would like to know if the configuration files you are using are vitl16_short.yaml and ssl_default_config.yaml, and if the batch size is 64x8? Also, have you used the official distributed module (submitit)? in my case， I used torch instead (because I couldn't run submitit), but the entire training process takes four to five days (the configuration files are vitl16_short.yaml and ssl_defaulteconfig.yaml, and the batch size is 64x8  ,devices are 8 A100 40G).
Can you provide a more detailed training process

**Comment by haoyudong-97** [comment]
Hi @mprzewie I'd like to see if you can obtain a similar top-1 performance on a single node with a larger number of OFFICIAL EPOCH LENGTH. Thanks!

**Comment by mprzewie** [comment]
Hi @ripperthejack , yes, I used the default configs you mention, except I think I managed to increase the per-GPU batch size to 128. I also used submitit. Not 100% sure about this, as I last worked with this code last year.

**Comment by mprzewie** [comment]
> Hi @mprzewie I'd like to see if you can obtain a similar top-1 performance on a single node with a larger number of OFFICIAL EPOCH LENGTH. Thanks!

Hi @haoyudong-97 , IIRC I did not increase this variable, so sadly cannot answer your question. I last worked with this codebase half a year ago, though, so I do not remember all the details.
